core/src/main/java/org/testcontainers/containers/GenericContainer.java

```diff
@@ -342,17 +319,63 @@ private void applyConfiguration(CreateContainerCmd createCommand) {
-                .map(entry -> new Link(entry.getValue().getContainerName(), entry.getKey()))
-                .collect(Collectors.toList())
-                .toArray(new Link[linkedContainers.size()]);
-        createCommand.withLinks(linksArray);
+        Set<Link> allLinks = new HashSet<>();
+        Set<String> allLinkedContainerNetworks = new HashSet<>();
+        for (Map.Entry<String, LinkableContainer> linkEntries : linkedContainers.entrySet()) {
+
+            String alias = linkEntries.getKey();
+            LinkableContainer linkableContainer = linkEntries.getValue();
+
+            Set<Link> links = dockerClient.listContainersCmd().exec().stream()
+                    .filter(container -> container.getNames()[0].endsWith(linkableContainer.getContainerName()))
+                    .map(container -> new Link(container.getNames()[0], alias))
+                    .collect(Collectors.toSet());
+            allLinks.addAll(links);
+
+            boolean linkableContainerIsRunning = dockerClient.listContainersCmd().exec().stream()
+                    .filter(container -> container.getNames()[0].endsWith(linkableContainer.getContainerName()))
+                    .map(com.github.dockerjava.api.model.Container::getId)
+                    .map(id -> dockerClient.inspectContainerCmd(id).exec())
+                    .anyMatch(linkableContainerInspectResponse -> linkableContainerInspectResponse.getState().getRunning());
+
+            if (!linkableContainerIsRunning) {
+                throw new ContainerLaunchException("Aborting attempt to link to container " +
+                        linkableContainer.getContainerName() +
+                        " as it is not running");
+            }
+
+            Set<String> linkedContainerNetworks = dockerClient.listContainersCmd().exec().stream()
+                    .filter(container -> container.getNames()[0].endsWith(linkableContainer.getContainerName()))
+                    .flatMap(container -> container.getNetworkSettings().getNetworks().keySet().stream())
+                    .distinct()
+                    .collect(Collectors.toSet());
+            allLinkedContainerNetworks.addAll(linkedContainerNetworks);
+        }
+
+        createCommand.withLinks(allLinks.toArray(new Link[allLinks.size()]));
+
+        allLinkedContainerNetworks.remove("bridge");
+        if (allLinkedContainerNetworks.size() > 1) {
+            logger().warn("Container needs to be on more than one custom network to link to other " +
+                            "containers - this is not currently supported. Required networks are: {}",
+                    allLinkedContainerNetworks);
+        }
+
+        Optional<String> networkForLinks = allLinkedContainerNetworks.stream().findFirst();
+        if (networkForLinks.isPresent()) {
+            logger().debug("Associating container with network: {}", networkForLinks.get());
+            createCommand.withNetworkMode(networkForLinks.get());
+        }
```

Comment:
As a general note - `applyConfiguration` is getting unwieldy and will be refactored in the near future!